src/config.rs

```diff
@@ -1042,4 +1189,90 @@ UNUSED = "unused"
 
         config.set("unknown-key", None).unwrap_err();
     }
+
+    #[rstest]
+    #[case("pinning-strategy", None, None)]
+    #[case("pinning-strategy", Some("semver".to_string()), Some(PinningStrategy::Semver))]
+    #[case("pinning-strategy", Some("no-pin".to_string()), Some(PinningStrategy::NoPin))]
+    #[case("pinning-strategy", Some("exact-version".to_string()), Some(PinningStrategy::ExactVersion))]
+    #[case("pinning-strategy", Some("latest-up".to_string()), Some(PinningStrategy::LatestUp))]
+    #[case("pinning-strategy", Some("major".to_string()), Some(PinningStrategy::Major))]
+    #[case("pinning-strategy", Some("minor".to_string()), Some(PinningStrategy::Minor))]
+    fn test_set_pinning_strategy(
+        #[case] key: &str,
+        #[case] value: Option<String>,
+        #[case] expected: Option<PinningStrategy>,
+    ) {
+        let mut config = Config::default();
+        config.set(key, value).unwrap();
+        assert_eq!(config.pinning_strategy, expected);
+    }
+
+    #[test]
+    fn test_version_constraints() {
```

Comment:
This is almost what I meant. I meant to create two sets, one with sets of versions (`[["1.0", "2.0"], ["1!1.2+3"]]`) and another with all the different strategies. Then combine the two (Cartesian product) and run all possible combinations. Then dump the result of that into a snapshot.

Reply:
Genius! Will change that!